### PR TITLE
Draft ax ordinal gp model

### PR DESCRIPTION
**Pull request by kamda-cyrial**
In this PR I,

1.) Create the `OrdinalGP` class which inherits from `VariationalGP`. I implement 2 methods `predict` and `get_config_options`
2.) Create a config file, `ax_ordinal_exploration_example.ini` to use for the implementation's testing in test_ax_intergration.py
3.) Implement some unit tests for the integration.
4.) add memorization test for `OrdinalGP` in test_variational_gp.py
5.) register the module in `aepsych/models/__init___.py`

### CODE REVIEW

File: aepsych/models/variational_gp.py [hunk 1]
```diff
@@ -113,4 +116,76 @@ def get_config_options(cls, config: Config, name: Optional[str] = None):
...
+    outcome_type = "ordinal"
+
+    def predict(self, xgrid):
+        with torch.no_grad():
+            post = self.posterior(xgrid)
+            fmean = promote_0d(post.mean.squeeze())
+            fvar = promote_0d(post.variance.squeeze())
+        return fmean, fvar
+
+    def predict_probs(self, xgrid):
```
Comment (crasanders) [inline aepsych/models/variational_gp.py:138]: Let's combine `predict` and `predict_probs` into a single `predict` method with a `probability_space` argument like in `BinaryClassificationGP`.

File: tests/test_ax_integration.py [hunk 2]
```diff
@@ -113,6 +118,127 @@ def test_generation_method(self):
...
+        config_file = "../configs/ax_ordinal_exploration_example.ini"
+        config_file = os.path.join(os.path.dirname(__file__), config_file)
+        cls.client.configure(config_file)
+        outcomes = []
+
+        # run a full synthetic experiment loop
+        while not cls.client.server.strat.finished:
+            try:
+                trial_config = cls.client.ask()
+                # if trial_config['is_finished']:
```
Comment (crasanders) [inline tests/test_ax_integration.py:173]: Commented-out code should be removed.

File: tests/test_ax_integration.py [hunk 3]
```diff
@@ -113,6 +118,127 @@ def test_generation_method(self):
...
+                #     print(cls.client.server.strat.strat._maybe_move_to_next_step())
+                #     continue
+                outcome = simulate_trial(trial_config=trial_config['config'])
+                cls.client.tell(config=trial_config['config'], outcome=outcome)
+                outcomes.append(outcome)
+            except Exception as e:
+                print(trial_config)
```
Comment (crasanders) [inline tests/test_ax_integration.py:180]: We shouldn't print anything during tests.

File: tests/test_ax_integration.py [hunk 4]
```diff
@@ -113,6 +118,127 @@ def test_generation_method(self):
...
 
+class AxOrdinalGPTestCase(unittest.TestCase):
+    """
+    Tests that the Ax integration works with an ordinal GP model.
+    """
+    @classmethod
+    def setUpClass(cls):
+
+        # Fix random seeds
+        np.random.seed(int(time.time()))
```
Comment (crasanders) [inline tests/test_ax_integration.py:129]: The random seed should be fixed to some fixed value like 0 or 1. The purpose of this is to be able to reproduce the same output every time you run the code.

File: tests/test_ax_integration.py [hunk 5]
```diff
@@ -113,6 +118,127 @@ def test_generation_method(self):
...
+                outcome = simulate_trial(trial_config=trial_config['config'])
+                cls.client.tell(config=trial_config['config'], outcome=outcome)
+                outcomes.append(outcome)
+            except Exception as e:
+                print(trial_config)
+                print(outcomes)
+                raise e
+            
+
+        # Add an extra tell to make sure manual tells and duplicate params
```
Comment (crasanders) [inline tests/test_ax_integration.py:185]: I don't understand this comment.

**Review by crasanders — CHANGES_REQUESTED** [review]
Looks good, but please see my comments.

**Review by crasanders — CHANGES_REQUESTED** [review]

File: tests/test_ax_integration.py [hunk 6]
```diff
@@ -113,6 +117,114 @@ def test_generation_method(self):
         self.assertEqual(n_sobol, correct_n_sobol)
         self.assertEqual(n_opt, correct_n_opt)
 
+class AxOrdinalGPTestCase(unittest.TestCase):
```
Comment (crasanders) [inline tests/test_ax_integration.py:120]: I wonder if instead of copy-pasting this test case for every model if we could make subclasses or pass in different arguments using [parameterized](https://pypi.org/project/parameterized/). But I am worried about making this test take too long to run, so I think by default we should only run the test using the continuous model (the current integration test).

File: tests/models/test_variational_gp.py [hunk 7]
```diff
@@ -61,5 +63,52 @@ def test_1d_classification(self):
...
+        X, y = make_classification(
+            n_samples=20,
+            n_features=5,
+            n_classes=cls.n_levels,
+            n_informative=3,
+            n_clusters_per_class=1,
+        )
+        cls.X, cls.y = torch.Tensor(X), torch.Tensor(y).reshape(-1, 1)
+
+    def test_ordinal_classification(self):
```
Comment (crasanders) [inline tests/models/test_variational_gp.py:81]: Since this test passes, does that assuage your worries about only get 0/4 predictions before?

**Review by crasanders — CHANGES_REQUESTED** [review]
This is close. See my comments and remove the changes to the notebooks from this PR.

File: tests/models/test_variational_gp.py (hunk 7 above)
Comment (kamda-cyrial) [reply tests/models/test_variational_gp.py:81]: yes, it does.

File: configs/ax_ordinal_exploration_example.ini [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+generator = SobolGenerator # The generator class used to generate new parameter values
+
+[opt_strat]
+generator = OptimizeAcqfGenerator # after sobol, do model-based active-learning
+min_total_tells = 3 # Finish the experiment after 3 total data points have been collected. Depending on how noisy
+                    # your problem is, you may need several dozen points per parameter to get an accurate model.
+acqf = qNoisyExpectedImprovement # The acquisition function to be used with the model. We recommend
+model = OrdinalGP
+
+# Configuration for the model
```
Comment (crasanders) [inline configs/ax_ordinal_exploration_example.ini:44]: Lines 44-48 should be deleted. We don't need this section because we are using `OrdinalGP`, not `ContinuousRegressionGP`.

File: tests/test_ax_integration.py (hunk 6 above)
Comment (crasanders) [reply tests/test_ax_integration.py:121]: I'd still like to see these integration tests generalized, but if that is not possible, you should at least put a `@unittest.skip` decorator like in the other PR.

File: aepsych/models/variational_gp.py [hunk 9]
```diff
@@ -115,3 +118,70 @@ def get_config_options(cls, config: Config, name: Optional[str] = None):
...
+    @classmethod
+    def get_config_options(cls, config: Config, name: Optional[str] = None):
+        options = super().get_config_options(config)
+
+        if options["likelihood"] is None:
+            options["likelihood"] = OrdinalLikelihood(n_levels=5)
+
+        dim = get_dim(config)
+
+        if config.getobj(cls.classname, "mean_covar_factory", fallback=None) is None:
```
Comment (mshvartsman) [inline aepsych/models/variational_gp.py:167]: `cls.__name__` would get you classname without hardcoding it above, I believe -- why do it this way?

File: aepsych/models/variational_gp.py [hunk 10]
```diff
@@ -115,3 +118,70 @@ def get_config_options(cls, config: Config, name: Optional[str] = None):
...
+    outcome_type = "ordinal"
+    classname = "OrdinalGP"
+
+    def predict(self, xgrid, probability_space=False):
+        with torch.no_grad():
+            post = self.posterior(xgrid)
+        fmean = post.mean.squeeze()
+        fvar = post.variance.squeeze()
+
+        if probability_space:
```
Comment (mshvartsman) [inline aepsych/models/variational_gp.py:137]: Is everything below here just copying `predict_probs` from `OrdinalGPModel`? Any way to make this less copy-pastey and ugly?

File: aepsych/utils.py [hunk 11]
```diff
@@ -277,3 +275,12 @@ def get_dim(config) -> int:
             )  # Choice parameters with n_choices < 3 add n_choices - 1 dims
 
     return dim
+
+
+def ignorable(func):
```
Comment (mshvartsman) [inline aepsych/utils.py:280]: Can you document what this wrapper does? And it looks like it should maybe be in some sort of `tests/utils.py` file or `tests/common.py` rather than in here?

File: tests/test_ax_integration.py [hunk 12]
```diff
@@ -57,12 +70,14 @@ def simulate_response(trial_params):
...
 
-        cls.config = Config(config_fnames=[config_file])
+        cls.config = Config(config_fnames=[cls.config_file])
 
+    @ignorable
     def tearDown(self):
         if self.client.server.db is not None:
             self.client.server.db.delete_db()
 
+    @ignorable
```
Comment (mshvartsman) [inline tests/test_ax_integration.py:80]: The fact that [a] you have to decorate every method in the class with this thing and [b] it's hard to understand what it does suggests to me that this is not the best approach here. Explicit is better than implicit, and clear is better than clever. If you can explain what you're trying to do here I wonder if we can come up with a better solution.

**Review by mshvartsman — CHANGES_REQUESTED** [review]

File: aepsych/models/variational_gp.py (hunk 10 above)
Comment (kamda-cyrial) [reply aepsych/models/variational_gp.py:137]: what are your thoughts on the newer approach

File: tests/test_ax_integration.py (hunk 12 above)
Comment (kamda-cyrial) [reply tests/test_ax_integration.py]: Indeed, it turns out I should have just raised the skip test exception to fulfil what I wanted. I just issued a fix.
